### PR TITLE
Upgrade to Terraform 0.12

### PR DESCRIPTION
**Pull request by robmorgan**
This PR upgrades `terraform-google-sql` to use Terraform 0.12. This is a backwards incompatible change.

- [x] Fix tests
- [x] Fix root example

### CODE REVIEW

File: modules/cloud-sql/variables.tf [hunk 1]
```diff
@@ -214,11 +244,11 @@ variable "resource_timeout" {
...
 # This can be used to make sure the module resources are created after other bootstrapping resources have been created.
 # For example:
-# dependencies = ["${google_service_networking_connection.private_vpc_connection.network}"]
+# dependencies = [google_service_networking_connection.private_vpc_connection.network]
 # ---------------------------------------------------------------------------------------------------------------------
 
 variable "dependencies" {
   description = "Create a dependency between the resources in this module to the interpolated values in this list (and thus the source resources). In other words, the resources in this module will now depend on the resources backing the values in this list such that those resources need to be created before the resources in this module, and the resources in this module need to be destroyed before the resources in the list."
-  type        = "list"
+  type        = list(any)
```
Comment (yorinasub17) [inline modules/cloud-sql/variables.tf]: This should be
```suggestion
  type        = list(string)
```

or `any` (without `list`). `list(any)` requires all the elements to have the same type, so it is much more error prone.

File: examples/mysql-private-ip/main.tf [hunk 2]
```diff
@@ -40,23 +40,23 @@ locals {
...
+  network       = google_compute_network.private_network.self_link
 }
 
 # Establish VPC network peering connection using the reserved address range
 resource "google_service_networking_connection" "private_vpc_connection" {
   provider                = "google-beta"
-  network                 = "${google_compute_network.private_network.self_link}"
+  network                 = google_compute_network.private_network.self_link
   service                 = "servicenetworking.googleapis.com"
   reserved_peering_ranges = ["${google_compute_global_address.private_ip_address.name}"]
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply examples/mysql-private-ip/main.tf]: @yorinasub17 No I did everything by hand, before realizing half-way I could have used the upgrade tool ;)

File: modules/cloud-sql/variables.tf [hunk 3]
```diff
@@ -42,12 +50,13 @@ variable "master_user_password" {
...
 variable "activation_policy" {
   description = "This specifies when the instance should be active. Can be either `ALWAYS`, `NEVER` or `ON_DEMAND`."
+  type        = string
   default     = "ALWAYS"
 }
 
 variable "authorized_networks" {
   description = "A list of authorized CIDR-formatted IP address ranges that can connect to this DB. Only applies to public IP instances."
-  type        = "list"
+  type        = list(any)
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply modules/cloud-sql/variables.tf]: ok let me try

File: modules/cloud-sql/main.tf [hunk 4]
```diff
@@ -192,61 +220,75 @@ resource "google_sql_database_instance" "failover_replica" {
...
 
     ip_configuration {
-      authorized_networks = ["${var.authorized_networks}"]
-      ipv4_enabled        = "${var.enable_public_internet_access}"
-      private_network     = "${var.private_network}"
-      require_ssl         = "${var.require_ssl}"
+      dynamic "authorized_networks" {
+        for_each = var.authorized_networks
+        content {
+          name  = authorized_networks.value.name
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply modules/cloud-sql/main.tf]: ah I have that before but wasn't sure if I needed it, so I ended up simplifying it.

File: modules/cloud-sql/variables.tf [hunk 5]
```diff
@@ -62,53 +71,61 @@ variable "authorized_networks" {
...
   description = "The collation for the default database. Example for MySQL databases: 'utf8_general_ci'."
-  default     = ""
+  type        = string
+  default     = null
 }
 
 variable "database_flags" {
   description = "List of Cloud SQL flags that are applied to the database server"
-  type        = "list"
+  type        = list(any)
```
[the remark this answers is not in the document]
Comment (robmorgan) [reply modules/cloud-sql/variables.tf]: @yorinasub17 they are both optional, but it's pointless to pass anything else or only pass one of them, should I make that change anyway?

File: modules/cloud-sql/variables.tf (hunk 1 above)
Comment (robmorgan) [reply modules/cloud-sql/variables.tf]: changed to `list(string)`

**Comment by robmorgan** [comment]
cc @vincentdeity

File: modules/cloud-sql/variables.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (robmorgan) [reply modules/cloud-sql/variables.tf]: I guess the `value` would need to be `any` as you can pass integers and strings etc: https://cloud.google.com/sql/docs/mysql/flags

**Comment by robmorgan** [comment]
I'm going to go ahead and release this. There's one minor open question about improving the types, but we can ship it in a patch release.

cc @yorinasub17

**Pull request merged**